Maximum Texture Resolution For Light Visualisers

mattigruener opened the pull request:
Hey John,

this is the code for the Gaffer side of things and depends on https://github.com/ImageEngine/cortex/pull/961. 

It's not ideal that we explicitly need to know about the "_maxResolution" naming convention here, hey? Also, I've jabbed that new attribute into StandardAttributes as I'm assuming that people here will want to apply it to lights that have already been published and come in via cache. 

I think the only lights that will currently use this new feature are lights in IERendering, though. If that's the case, would you rather we only make changes to the visualiser here and handle the attributes setup in IERendering? That would have the downside that published lights can't be easily modified, but then we don't have to deal with it here.
If we opt to add the attribute to StandardAttributes, is it worth looking at "visualiser:scale" again and add it to the same section? Maybe with a tooltip saying that it's only for lights currently? It feels to me like that section could be useful to have. I was thinking about making our gobo visualiser respect the texture resolution attribute to figure out the grid size on which the osl network gets evaluated, for example...

Cheerio,

Matti.

johnhaddon commented on include/GafferSceneUI/StandardLightVisualiser.h:
```diff
@@ -74,7 +74,7 @@ class GAFFERSCENEUI_API StandardLightVisualiser : public LightVisualiser
 		static IECoreGL::ConstRenderablePtr distantRays();
 		static IECoreGL::ConstRenderablePtr spotlightCone( float innerAngle, float outerAngle, float lensRadius );
-		static IECoreGL::ConstRenderablePtr environmentSphere( const Imath::Color3f &color, const std::string &textureFileName );
+		static IECoreGL::ConstRenderablePtr environmentSphere( const Imath::Color3f &color, const std::string &textureFileName, const IECore::IntData *maxTextureResolution );
```
As per the Cortex review, let's use an `int` rather than `IntData`.

mattigruener replied:
Addressed in 8c2e916.

johnhaddon commented on src/GafferScene/StandardAttributes.cpp:
```diff
@@ -69,6 +69,10 @@ StandardAttributes::StandardAttributes( const std::string &name )
 
 	attributes->addChild( new Gaffer::NameValuePlug( "filteredLights", new IECore::StringData( "" ), false, "filteredLights" ) );
 
+	// visualisers
```
I think this belongs on OpenGLAttributes rather than StandardAttributes - that's where everything else specific to the viewer is. The general idea with StandardAttributes is that it contains attributes that could be applicable to _any_ renderer.

mattigruener replied:
Addressed in 20367a5. I've also renamed the section (now "Light Visualisers") and pefixed the attribute name with "gl:". Additionally, the default value is 512 now as per what the `GoboVisualiser` used.

johnhaddon commented on src/GafferSceneUI/StandardLightVisualiser.cpp:
```diff
@@ -571,8 +572,14 @@ IECoreGL::ConstRenderablePtr StandardLightVisualiser::environmentSphere( const I
+
+	if( textureMaxResolution )
+	{
+		parameters->members()["mapSampler_maxResolution"] = textureMaxResolution->copy();
```
What's the failure mode for this when used with a Cortex without support? Is it just a warning, or does it break the drawing in some way? Just wondering if we can merge in anticipation of the Cortex update or if it should wait.

mattigruener replied:
I couldn't see any problems. Didn't even see a warning.

mattigruener commented:
Hey John,

your feedback is addressed inline.

I've also added the following:

- `GoboVisualiser` now respects that new attribute (57afc07). I cleaned up a violation of our conventions, while I was looking at that file (d7f7c9a).
- Since I was dealing with visualisation and OSL evaluation already, I took a stab at extending those features. I'm happy to push this into a separate PR, if you think this is too far from mergeable. The functionality is useful, I guess? I'm listing my changes below.
  - The code that does the OSL evaluation is now in `StandardLightVisualiser` which required a linking update as described in ce7c1e3. That should be ok? `GafferSceneUI` now depends on `GafferOSL`.
  - Small fix to invert v for visualised grids in 5b9644d.
  - As per a comment in `ArnoldLight`, I've changed that class to rely on an internal shader rather than managing `ShaderNetworks` itself in 75a5f0b.
  - All of the above allow the changes I've made in 3f0e0f2 which allows us to visualise osl driven aiQuadlights in Gaffer's viewport.

The last bits are definitely outside of the original scope of this PR. Just let me know if you feel like we want to merge the initial work independently and need to talk more about the area light visualisation. Since you said it's a priority for our global efforts, I thought I'd take a look while I was at it...

Leaving the label for now as I need to clean up the history after you looked at the changes.

Thanks :)

Matti.

mattigruener commented:
@medubelko, things that we potentially want to document:

`OpenGLAttributes` now has an attribute that limits the texture resolution for images used as part of light visualisations. As far as public lights go, it's supported by Appleseed's environment light and Arnold's gobos. Arnold's "quad_light" is added to this list if John likes my changes to ArnoldLight etc. Will get back to you regarding that.

mattigruener commented:
Sorry, I guess I hadn't re-requested a review. Still getting used to this new system.

mattigruener commented:
Hey John,

thanks for the review :) I've now squashed all changes into just the two remaining commits. For the other commits I'll put up a separate PR and respond to your questions there.

Thanks,

Matti.
